Implementation of feeds page

ddey2 opened the pull request:
Try to create a feed from UI

Steps to reproduce:
1. Go to Feeds menu on the left side pannel
2. Create a feed
3. Edit a Feed
4. Delete a Feed
<img width="1715" alt="Screenshot 2024-05-22 at 8 13 36 PM" src="https://github.com/clowder-framework/clowder2/assets/22405928/ef3df834-76d8-49c0-bc5f-f76a1ffa80f8">
<img width="1714" alt="Screenshot 2024-05-22 at 8 07 03 PM" src="https://github.com/clowder-framework/clowder2/assets/22405928/3423920a-7bc0-45af-9737-35d79ae99ba9">
<img width="1669" alt="Screenshot 2024-05-22 at 8 07 12 PM" src="https://github.com/clowder-framework/clowder2/assets/22405928/30b30c12-8d7e-4264-9a8c-a2b685183bda">
<img width="1255" alt="Screenshot 2024-05-22 at 8 07 36 PM" src="https://github.com/clowder-framework/clowder2/assets/22405928/7375a1ab-6ff2-42b4-ba56-a9986b75beba">

ddey2 commented:
need to work on small fixes like page doesn't reload after you edit a feed. working on that but rest is good for review and any feedback.

ddey2 commented:
Feeds page
<img width="1697" alt="Screenshot 2024-05-30 at 10 42 16 AM" src="https://github.com/clowder-framework/clowder2/assets/22405928/cd153477-95d2-474a-99f1-816c3f3215a9">


Note: the background css of top bar changes when you open a modal. Looking for a fix for that.

longshuicy reviewed (commented):
<img width="1164" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/ae450ee9-754a-489a-8ac3-f2d4a0c46b2c">

1. What does index mean? 
2. Is it possible you could use the styled widgets to avoid the mess up of thebackground style. Feel free to address it in another task though

<img width="1776" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/6872ebe2-44d6-42ab-a91f-04c7d283e341">

When I have some unfilled field, then I click save, it becomes a blank page. 
If I fill every field, then click save, it is not immediately reflected. I have to refresh the page.

<img width="1075" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/4dca85c2-cfff-46db-8e59-f8c70980e3c1">
Could we group these two button/icons. Or at least match the table header column, right now it's missing one column

ddey2 commented on frontend/src/components/listeners/Feeds.tsx:
```diff
@@ -0,0 +1,282 @@
+											{/*	>*/}
+											{/*		{feed.name}*/}
+											{/*	</MuiLink>*/}
+											{/*</TableCell>*/}
```
done

ddey2 commented on frontend/src/components/listeners/CreateFeedModal.tsx:
```diff
@@ -0,0 +1,81 @@
+				{/*						<span>{listener.name}*/}
+				{/*						</span>*/}
+				{/*						)})): <></>}*/}
+				{/*</List>*/}
```
done

ddey2 commented on frontend/src/components/listeners/CreateFeedModal.tsx:
```diff
@@ -0,0 +1,81 @@
+	const schema = feedSchema["schema"];
+
+	const saveFeed = (formData: FormData) => {
+		console.log(formData);
```
done

ddey2 commented on frontend/src/components/listeners/Feeds.tsx:
```diff
@@ -0,0 +1,282 @@
+	// component did mount
+
+	useEffect(() => {
+		listFeeds("", 0, limit);
```
When a feed is created/edited, I am updating the feed state. Hence I want to reload the page. But I guess I can use the curPageNum. Will do that. Thanks for pointing out.

ddey2 commented:
> <img alt="image" width="1164" src="https://private-user-images.githubusercontent.com/13950475/335373486-ae450ee9-754a-489a-8ac3-f2d4a0c46b2c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc2OTExMTAsIm5iZiI6MTcxNzY5MDgxMCwicGF0aCI6Ii8xMzk1MDQ3NS8zMzUzNzM0ODYtYWU0NTBlZTktNzU0YS00ODlhLThhYzMtZjJkNGEwYzQ2YjJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA2VDE2MjAxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIwZDc2OTk3YjU0ODA4YjEyYTNjZWRlOTY4MTQwNDczNDU1MzkyZGU0Yjk4ZWY3OGJjYzNhYWQyN2IzODc2YTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OSmCgI_KG2_F7T3YYZfPQZeXp0hvJP8bYaIrk9nUni0">
> 1. What does index mean?
> 2. Is it possible you could use the styled widgets to avoid the mess up of thebackground style. Feel free to address it in another task though
> 
> <img alt="image" width="1776" src="https://private-user-images.githubusercontent.com/13950475/335374613-6872ebe2-44d6-42ab-a91f-04c7d283e341.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc2OTExMTAsIm5iZiI6MTcxNzY5MDgxMCwicGF0aCI6Ii8xMzk1MDQ3NS8zMzUzNzQ2MTMtNjg3MmViZTItNDRkNi00MmFiLWE5MWYtMDRjN2QyODNlMzQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA2VDE2MjAxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUzOWRlY2I2ZmJkYmQ3ODE4YjE1YjExODEwOGM4NTg5NGQyMTRmODc4NDc3N2U4MmNiMGNmMjE2OThhYWNmYzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.uUC04ZaafCuTopY0hI3YMAoXu6AS_6OCys6rzO1avR4">
> When I have some unfilled field, then I click save, it becomes a blank page. If I fill every field, then click save, it is not immediately reflected. I have to refresh the page.
> 
> <img alt="image" width="1075" src="https://private-user-images.githubusercontent.com/13950475/335375598-4dca85c2-cfff-46db-8e59-f8c70980e3c1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc2OTExMTAsIm5iZiI6MTcxNzY5MDgxMCwicGF0aCI6Ii8xMzk1MDQ3NS8zMzUzNzU1OTgtNGRjYTg1YzItY2ZmZi00NmRiLThlNTktZjhjNzA5ODBlM2MxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA2VDE2MjAxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIzMDgxMmI1MmU4NDk0MzBlN2Y2Y2YyNGY5OTMwYTViYWVlOTY2YWQyMjg3MGVkM2FmMjgzNzFiNjFmNTQ0NDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.aIwZS2DF4BVX6CVe5N23TruFvPhgeJ0agmADlauilLs"> Could we group these two button/icons. Or at least match the table header column, right now it's missing one column

I have made each sub property of feedSchema as required. It should not let you submit the form unless you fill in each details. Please check.

I ha
<img width="1617" alt="Screenshot 2024-06-06 at 2 02 07 PM" src="https://github.com/clowder-framework/clowder2/assets/22405928/32bf83e6-e0a7-42d5-be5b-33968e0d0cf4">
ve grouped the button together

ddey2 commented on backend/app/routers/feeds.py:
```diff
@@ -71,28 +73,80 @@ async def save_feed(
 
 
-@router.get("", response_model=List[FeedOut])
+@router.post("/{feed_id}", response_model=FeedOut)
```
done

ddey2 commented on backend/app/routers/feeds.py:
```diff
@@ -71,28 +73,80 @@ async def save_feed(
+    if searchTerm is not None:
+        criteria_list.append(
+            Or(
+                RegEx(field=FeedDB.name, pattern=searchTerm),
```
Thanks for noticing this. Added.

ddey2 commented on backend/app/routers/feeds.py:
```diff
@@ -71,28 +73,80 @@ async def save_feed(
+    """
+    feed = await FeedDB.get(PydanticObjectId(feed_id))
+    if feed:
+        # TODO: Refactor this with permissions checks etc.
```
Took care of permission in this PR

ddey2 commented:
Used rjsf/mui instead of rjsf/metrail-ui
Updated the component name and added dependent components in package.json
Updated several places in code where react json form is used

lmarini reviewed (changes requested):
Now that we use a single elasticsearch index, could we remove the `index` field from both the UI as well as the model?

longshuicy commented on frontend/src/components/datasets/CreateLicenseModal.tsx:
```diff
@@ -12,8 +12,9 @@ import {
-import { V2 } from "../../openapi"; // Import LicenseOut type
+import Form from "@rjsf/mui";
+import { V2 } from "../../openapi";
+import validator from "@rjsf/validator-ajv8"; // Import LicenseOut type
```
Maybe wrong inline comment "// Import LicenseOut type"?

ddey2 replied:
No idea how this was added. But will remove that.

longshuicy commented on frontend/src/components/listeners/FeedEntry.tsx:
```diff
@@ -0,0 +1,223 @@
+		listListeners();
+	}, []);
+
+	useEffect(() => {}, [feed]);
```
Is this empty useEffect hook needed?

ddey2 replied:
removed

longshuicy commented on frontend/package.json:
```diff
@@ -43,8 +43,10 @@
-		"@rjsf/core": "^3.2.1",
-		"@rjsf/material-ui": "^3.2.1",
+		"@rjsf/core": "^5.18.4",
+		"@rjsf/utils": "^5.18.4",
```
is @rjsf/utils used?

ddey2 replied:
Yes, it's needed. It's a dependency for rjsf/mui to work.

ddey2 replied:
Refer to Prerequisites in https://www.npmjs.com/package/@rjsf/mui

longshuicy commented on frontend/src/schema/feedSchema.json:
```diff
@@ -0,0 +1,91 @@
+								"operator": {
+									"type": "string",
+									"title": "Operator",
+									"enum": ["==", "<", ">", "!="]
```
It would be nice to add "regex" support for the future so we can match even fine gridded. Is there a separate task for that?

ddey2 replied:
For now, I think we wanted it as a fixed list of options , as per discussion with @lmarini
However I created a separate task #1102

longshuicy commented on backend/app/heartbeat_listener_sync.py:
```diff
@@ -86,7 +86,6 @@ def callback(ch, method, properties, body):
                 new_feed = FeedDB(
                     name=extractor_name,
```
You can also try add the description extractor_info["description"]. Something like this:
```
extractor_description = extractor_name
if "description" in extractor_info:
  extractor_description = extractor_info["description"]
```

ddey2 replied:
added

longshuicy reviewed (changes requested):
Tested creating specific automatic extracting feed and it works as expected. 
Delete feeds also works.

I noticed some glitch with edit the feeds though:
<img width="1300" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/19f0e623-9b5c-4c1d-a55b-9409678d66e2">
In **edit** mode, seems like I cannot select an extractor. But I can select when in "create" mode.

longshuicy commented:
@ddey2 did you push your change? Am I missing anything since I didn't see any updates since my last comment :-(

ddey2 commented:
> @ddey2 did you push your change? Am I missing anything since I didn't see any updates since my last comment :-(

My bad. I made the commit and forgot to push it to remote. Now check.

longshuicy reviewed (commented):
Could you reproduce these:
1. I can set criteria and extractor as empty? I recall you forbids that last time I reviewed
2. After delete it didn't go away on the frontend

Screenshots attached. 
<img width="1025" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/661088a0-c70b-4b31-be53-091fd20fedf5">
<img width="895" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/b29fbc07-26fc-4662-a79a-71b0190a8574">

ddey2 commented:
> Could you reproduce these:
> 
> 1. I can set criteria and extractor as empty? I recall you forbids that last time I reviewed
> 2. After delete it didn't go away on the frontend
> 
> Screenshots attached. <img alt="image" width="1025" src="https://private-user-images.githubusercontent.com/13950475/339113347-661088a0-c70b-4b31-be53-091fd20fedf5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyMjM0OTYsIm5iZiI6MTcxODIyMzE5NiwicGF0aCI6Ii8xMzk1MDQ3NS8zMzkxMTMzNDctNjYxMDg4YTAtYzcwYi00YjMxLWJlNTMtMDkxZmQyMGZlZGY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEyVDIwMTMxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFiZTMzNzUyMzg0ZDgwYjBlZDYyODNjMmJjZmY2MTcyNGIyYmQ0ZGM1ZWI1ZTJjYTgxOTdiMDZhYjhlYmIxYzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZIsiSdAzeon071I_4pmZY96dnAUtMGyQJOvCdcDm2bk"> <img alt="image" width="895" src="https://private-user-images.githubusercontent.com/13950475/339113759-b29fbc07-26fc-4662-a79a-71b0190a8574.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyMjM0OTYsIm5iZiI6MTcxODIyMzE5NiwicGF0aCI6Ii8xMzk1MDQ3NS8zMzkxMTM3NTktYjI5ZmJjMDctMjZmYy00NjYyLWE3OWEtNzFiMDE5MGE4NTc0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEyVDIwMTMxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIzN2MwODViNmEwZWVmNzhlNWFlMTQ4NzMwZjk0MTJhNTg2NDY1MTA3ZmU5M2ViM2ZiZGVhY2VhYmFhZDM1NDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.b3qib_mMavtGhk_mxbUXO9xgbLjT6aPhocyz5Uintc0">

Thanks so much for noticing these out. Fixed now.

1 is related to making change to rsjf/mui. Some extra properties are added to make sure those aren't empty.

longshuicy reviewed (approved):
Everything works as expected now. Approved.
